{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "B2-1FFdTtCmA"
      },
      "source": [
        "# 1. Mount gDrive that has the folder containing:\n",
        "- **Genomic data:** folder containing the 120 genome fasta files\n",
        "- **Phenotypic data:** Metadata.tsv file containing information about biofilm-formation capability of each isolate\n",
        "(0 --> non-biofilm forming, 1 --> biofilm-forming)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4G3GElwnYaeA",
        "outputId": "d26eb1e5-b990-4871-bf9c-8dccd9c0987c"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /gdrive\n"
          ]
        }
      ],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/gdrive')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "hK1-FojEt_oA"
      },
      "source": [
        "# 2. Reading the kmers generated from the genes of interest and reading the isolate metadata.\n",
        "\n",
        "- The kmers generated from the genes of interest are stored in the *all_gene_kmers.txt* file (Column 1: 31-mer, Column 2: non-zero frequency).\n",
        "\n",
        "- The isolate IDs and labels can be extracted from the *Metadata.tsv* file (Column 1: Isolate ID, Column 2: Isolate label).\n",
        "\n",
        "The kmers will be used as column names in the kmer presence/absence matrix.\n",
        "The isolate metadata will be needed to label the rows of the matrix and to add the 'Label' column at the end of the matrix.\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%cd /gdrive\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "FI8fVNTJD2X4",
        "outputId": "23d44a06-f17e-4445-b2b1-958858923783"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/gdrive\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "import numpy as np\n",
        "import pandas as pd"
      ],
      "metadata": {
        "id": "3U2UCmoGD6ZL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "g3YZDE9EVR7H",
        "outputId": "b2885827-e0ec-4425-b069-b06d0da445fd"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/gdrive\n",
            "Isolate metadata:  {'CA_001': '1', 'CA_002': '1', 'CA_003': '0', 'CA_005': '1', 'CA_007': '0', 'CA_008': '1', 'CA_009': '1', 'CA_010': '1', 'CA_011': '1', 'CA_012': '1', 'CA_013': '1', 'CA_014': '1', 'CA_015': '1', 'CA_016': '1', 'CA_017': '1', 'CA_018': '1', 'CA_019': '1', 'CA_020': '1', 'CA_021': '1', 'CA_022': '0', 'CA_023': '1', 'CA_024': '0', 'CA_025': '0', 'CA_026': '1', 'CA_027': '1', 'CA_028': '0', 'CA_029': '1', 'CA_030': '1', 'CA_031': '1', 'CA_032': '0', 'CA_033': '1', 'CA_034': '0', 'CA_035': '0', 'CA_036': '1', 'CA_038': '1', 'CA_039': '0', 'CA_040': '1', 'CA_041': '1', 'CA_042': '0', 'CA_043': '1', 'CA_044': '0', 'CA_045': '1', 'CA_046': '0', 'CA_047': '0', 'CA_048': '1', 'CA_049': '0', 'CA_050': '0', 'CA_052': '1', 'CA_053': '1', 'CA_054': '1', 'CA_055': '1', 'CA_057': '1', 'CA_058': '1', 'CA_059': '1', 'CA_060': '1', 'CA_062': '0', 'CA_063': '0', 'CA_064': '1', 'CA_066': '0', 'CA_068': '0', 'CA_070': '1', 'CA_074': '0', 'CA_076': '1', 'CA_078': '1', 'CA_086': '1', 'CA_088': '1', 'CA_089': '1', 'CA_090': '1', 'CA_091': '1', 'CA_092': '1', 'CA_093': '1', 'CA_094': '1', 'CA_095': '1', 'CA_096': '1', 'CA_101': '1', 'CA_104': '1', 'CA_105': '1', 'CA_107': '1', 'CA_108': '1', 'CA_112': '1', 'CA_115': '1', 'CA_117': '1', 'CA_118': '1', 'CA_119': '1', 'CA_120': '1', 'CA_127': '1', 'CA_128': '1', 'CA_129': '1', 'CA_130': '1', 'CA_131': '1', 'CA_132': '1', 'CA_138': '1', 'CA_140': '1', 'CA_141': '1', 'CA_142': '1', 'CA_143': '1', 'CA_146': '0', 'CA_148': '1', 'CA_150': '0', 'CA_151': '1', 'CA_152': '0', 'CA_153': '1', 'CA_154': '1', 'CA_155': '0', 'CA_156': '0', 'CA_158': '1', 'CA_162': '1', 'CA_163': '1', 'CA_164': '0', 'CA_165': '1', 'CA_167': '1', 'CA_168': '1', 'CA_177': '1', 'CA_185': '1', 'CA_186': '1', 'CA_187': '0', 'CA_188': '1', 'CA_190': '1', 'CA_198': '1', 'CA_200': '1'}\n",
            "Number of kmers generated by genes of interest (expected output: 22,435): 22435\n"
          ]
        }
      ],
      "source": [
        "import os\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "\n",
        "# Change working directory to the drive that was mounted\n",
        "%cd /gdrive\n",
        "\n",
        "# Get the path of the folder containing the Genomic and Phenotypic data\n",
        "directory_path = os.path.join(os.getcwd(), \"MyDrive/ML Proj data/RF\")\n",
        "folder_content = os.listdir(directory_path)\n",
        "\n",
        "kmers = [] # Will store the kmer names to be used as column names when building the matrix\n",
        "labels = {} # Key: Isolate ID, Value: Isolate label\n",
        "\n",
        "for file in folder_content:\n",
        "\n",
        "  # Storing isolate metadata from the Metadata.tsv file\n",
        "  if 'Metadata.tsv' in file:\n",
        "    file_path = os.path.join(directory_path, file)\n",
        "    file = open(file_path, 'r')\n",
        "    for line in file:\n",
        "      line = line.rstrip().split('\\t')\n",
        "      labels[line[0]] = line[1]\n",
        "\n",
        "  # Storing kmers from all_gene_kmers.txt file\n",
        "  if 'all_gene_kmers.txt' in file:\n",
        "        file_path = os.path.join(directory_path, file)\n",
        "        file = open(file_path, 'r')\n",
        "        for line in file.readlines():\n",
        "          line = line.rstrip().split()\n",
        "          kmers.append(line[0])\n",
        "\n",
        "# To check that the isolate metadata was correctly stored\n",
        "print('Isolate metadata: ', labels)\n",
        "\n",
        "# To check that all the kmers were read\n",
        "print('Number of kmers generated by genes of interest (expected output: 22,435):', len(kmers))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WdyeLvn8w9BR"
      },
      "source": [
        "# 3. Building Kmer presence/absence dataframe out of the 22,435 kmers of the genes of interest and 120 genomes to be investigated.\n",
        "\n",
        "- The kmers will form the matrix columns\n",
        "- The isolates will for the matrix rows\n",
        "- The cells of the matrix will be filled with 0s and 1s based on the absence/presence of a certain kmer in a certain isolate genome, respectively."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "fhFmioPhJLw7",
        "outputId": "ddcba75a-518b-49d3-9788-1013123e722e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Directory content: ['CA_001.txt', 'CA_002.txt', 'CA_003.txt', 'CA_005.txt', 'CA_007.txt', 'CA_008.txt', 'CA_009.txt', 'CA_010.txt', 'CA_011.txt', 'CA_012.txt', 'CA_013.txt', 'CA_014.txt', 'CA_015.txt', 'CA_016.txt', 'CA_017.txt', 'CA_018.txt', 'CA_019.txt', 'CA_020.txt', 'CA_021.txt', 'CA_022.txt', 'CA_023.txt', 'CA_024.txt', 'CA_025.txt', 'CA_026.txt', 'CA_027.txt', 'CA_028.txt', 'CA_029.txt', 'CA_030.txt', 'CA_031.txt', 'CA_032.txt', 'CA_033.txt', 'CA_034.txt', 'CA_035.txt', 'CA_036.txt', 'CA_038.txt', 'CA_039.txt', 'CA_040.txt', 'CA_041.txt', 'CA_042.txt', 'CA_043.txt', 'CA_044.txt', 'CA_045.txt', 'CA_046.txt', 'CA_047.txt', 'CA_048.txt', 'CA_049.txt', 'CA_050.txt', 'CA_052.txt', 'CA_053.txt', 'CA_054.txt', 'CA_055.txt', 'CA_057.txt', 'CA_058.txt', 'CA_059.txt', 'CA_060.txt', 'CA_062.txt', 'CA_063.txt', 'CA_064.txt', 'CA_066.txt', 'CA_068.txt', 'CA_070.txt', 'CA_074.txt', 'CA_076.txt', 'CA_078.txt', 'CA_086.txt', 'CA_088.txt', 'CA_089.txt', 'CA_090.txt', 'CA_091.txt', 'CA_092.txt', 'CA_093.txt', 'CA_094.txt', 'CA_095.txt', 'CA_096.txt', 'CA_101.txt', 'CA_104.txt', 'CA_105.txt', 'CA_107.txt', 'CA_108.txt', 'CA_112.txt', 'CA_115.txt', 'CA_117.txt', 'CA_118.txt', 'CA_119.txt', 'CA_120.txt', 'CA_127.txt', 'CA_128.txt', 'CA_129.txt', 'CA_130.txt', 'CA_131.txt', 'CA_132.txt', 'CA_138.txt', 'CA_140.txt', 'CA_141.txt', 'CA_142.txt', 'CA_143.txt', 'CA_146.txt', 'CA_148.txt', 'CA_150.txt', 'CA_151.txt', 'CA_152.txt', 'CA_153.txt', 'CA_154.txt', 'CA_155.txt', 'CA_156.txt', 'CA_158.txt', 'CA_162.txt', 'CA_163.txt', 'CA_164.txt', 'CA_165.txt', 'CA_167.txt', 'CA_168.txt', 'CA_177.txt', 'CA_185.txt', 'CA_186.txt', 'CA_187.txt', 'CA_188.txt', 'CA_190.txt', 'CA_198.txt', 'CA_200.txt']\n",
            "Number\t Isolate ID\t Number of kmers found in isolate\n",
            "1\t  CA_001\t\t\t715\n",
            "2\t  CA_002\t\t\t379\n",
            "3\t  CA_003\t\t\t868\n",
            "4\t  CA_005\t\t\t588\n",
            "5\t  CA_007\t\t\t394\n",
            "6\t  CA_008\t\t\t586\n",
            "7\t  CA_009\t\t\t374\n",
            "8\t  CA_010\t\t\t702\n",
            "9\t  CA_011\t\t\t379\n",
            "10\t  CA_012\t\t\t395\n",
            "11\t  CA_013\t\t\t828\n",
            "12\t  CA_014\t\t\t702\n",
            "13\t  CA_015\t\t\t461\n",
            "14\t  CA_016\t\t\t533\n",
            "15\t  CA_017\t\t\t369\n",
            "16\t  CA_018\t\t\t520\n",
            "17\t  CA_019\t\t\t702\n",
            "18\t  CA_020\t\t\t680\n",
            "19\t  CA_021\t\t\t309\n",
            "20\t  CA_022\t\t\t739\n",
            "21\t  CA_023\t\t\t702\n",
            "22\t  CA_024\t\t\t492\n",
            "23\t  CA_025\t\t\t491\n",
            "24\t  CA_026\t\t\t507\n",
            "25\t  CA_027\t\t\t481\n",
            "26\t  CA_028\t\t\t660\n",
            "27\t  CA_029\t\t\t374\n",
            "28\t  CA_030\t\t\t575\n",
            "29\t  CA_031\t\t\t403\n",
            "30\t  CA_032\t\t\t441\n",
            "31\t  CA_033\t\t\t663\n",
            "32\t  CA_034\t\t\t367\n",
            "33\t  CA_035\t\t\t512\n",
            "34\t  CA_036\t\t\t397\n",
            "35\t  CA_038\t\t\t510\n",
            "36\t  CA_039\t\t\t486\n",
            "37\t  CA_040\t\t\t508\n",
            "38\t  CA_041\t\t\t437\n",
            "39\t  CA_042\t\t\t477\n",
            "40\t  CA_043\t\t\t531\n",
            "41\t  CA_044\t\t\t453\n",
            "42\t  CA_045\t\t\t466\n",
            "43\t  CA_046\t\t\t771\n",
            "44\t  CA_047\t\t\t432\n",
            "45\t  CA_048\t\t\t741\n",
            "46\t  CA_049\t\t\t588\n",
            "47\t  CA_050\t\t\t762\n",
            "48\t  CA_052\t\t\t409\n",
            "49\t  CA_053\t\t\t533\n",
            "50\t  CA_054\t\t\t786\n",
            "51\t  CA_055\t\t\t461\n",
            "52\t  CA_057\t\t\t390\n",
            "53\t  CA_058\t\t\t519\n",
            "54\t  CA_059\t\t\t731\n",
            "55\t  CA_060\t\t\t143\n",
            "56\t  CA_062\t\t\t406\n",
            "57\t  CA_063\t\t\t702\n",
            "58\t  CA_064\t\t\t421\n",
            "59\t  CA_066\t\t\t1375\n",
            "60\t  CA_068\t\t\t3794\n",
            "61\t  CA_070\t\t\t3421\n",
            "62\t  CA_074\t\t\t771\n",
            "63\t  CA_076\t\t\t648\n",
            "64\t  CA_078\t\t\t819\n",
            "65\t  CA_086\t\t\t1382\n",
            "66\t  CA_088\t\t\t380\n",
            "67\t  CA_089\t\t\t472\n",
            "68\t  CA_090\t\t\t394\n",
            "69\t  CA_091\t\t\t499\n",
            "70\t  CA_092\t\t\t739\n",
            "71\t  CA_093\t\t\t354\n",
            "72\t  CA_094\t\t\t432\n",
            "73\t  CA_095\t\t\t1169\n",
            "74\t  CA_096\t\t\t395\n",
            "75\t  CA_101\t\t\t149\n",
            "76\t  CA_104\t\t\t263\n",
            "77\t  CA_105\t\t\t614\n",
            "78\t  CA_107\t\t\t178\n",
            "79\t  CA_108\t\t\t643\n",
            "80\t  CA_112\t\t\t750\n",
            "81\t  CA_115\t\t\t696\n",
            "82\t  CA_117\t\t\t344\n",
            "83\t  CA_118\t\t\t577\n",
            "84\t  CA_119\t\t\t376\n",
            "85\t  CA_120\t\t\t376\n",
            "86\t  CA_127\t\t\t588\n",
            "87\t  CA_128\t\t\t713\n",
            "88\t  CA_129\t\t\t514\n",
            "89\t  CA_130\t\t\t388\n",
            "90\t  CA_131\t\t\t469\n",
            "91\t  CA_132\t\t\t739\n",
            "92\t  CA_138\t\t\t440\n",
            "93\t  CA_140\t\t\t434\n",
            "94\t  CA_141\t\t\t457\n",
            "95\t  CA_142\t\t\t549\n",
            "96\t  CA_143\t\t\t397\n",
            "97\t  CA_146\t\t\t736\n",
            "98\t  CA_148\t\t\t446\n",
            "99\t  CA_150\t\t\t354\n",
            "100\t  CA_151\t\t\t731\n",
            "101\t  CA_152\t\t\t429\n",
            "102\t  CA_153\t\t\t434\n",
            "103\t  CA_154\t\t\t572\n",
            "104\t  CA_155\t\t\t430\n",
            "105\t  CA_156\t\t\t329\n",
            "106\t  CA_158\t\t\t731\n",
            "107\t  CA_162\t\t\t706\n",
            "108\t  CA_163\t\t\t407\n",
            "109\t  CA_164\t\t\t588\n",
            "110\t  CA_165\t\t\t634\n",
            "111\t  CA_167\t\t\t431\n",
            "112\t  CA_168\t\t\t414\n",
            "113\t  CA_177\t\t\t395\n",
            "114\t  CA_185\t\t\t406\n",
            "115\t  CA_186\t\t\t396\n",
            "116\t  CA_187\t\t\t362\n",
            "117\t  CA_188\t\t\t314\n",
            "118\t  CA_190\t\t\t470\n",
            "119\t  CA_198\t\t\t533\n",
            "120\t  CA_200\t\t\t533\n",
            "Are there any null values in the dataframe? False\n",
            "Final Kmer presence/absence matrix:\n",
            "         AAAACAACGCTAATCAAGCGAACAATCAAGC  AAACAACGCTAATCAAGCGAACAATCAAGCA  \\\n",
            "CA_001                              0.0                              0.0   \n",
            "CA_002                              0.0                              0.0   \n",
            "CA_003                              0.0                              0.0   \n",
            "CA_005                              0.0                              0.0   \n",
            "CA_007                              0.0                              0.0   \n",
            "\n",
            "        AAATTATTAAAGTTTAAAGAATAACCACCAT  AAAGATTTTTGTAGTTCTTCATCTTGGTTAA  \\\n",
            "CA_001                              0.0                              0.0   \n",
            "CA_002                              0.0                              0.0   \n",
            "CA_003                              0.0                              0.0   \n",
            "CA_005                              0.0                              0.0   \n",
            "CA_007                              0.0                              0.0   \n",
            "\n",
            "        AACAACGCTAATCAAGCGAACAATCAAGCAC  AACCAAGATGAAGAACTACAAAAATCTTTAG  \\\n",
            "CA_001                              0.0                              0.0   \n",
            "CA_002                              0.0                              0.0   \n",
            "CA_003                              0.0                              0.0   \n",
            "CA_005                              0.0                              0.0   \n",
            "CA_007                              0.0                              0.0   \n",
            "\n",
            "        AATCCGAATCGCTATCTGAATCTGAATCGCT  AATCTGAATCGCTATCCGAGTCTGAGTCGCT  \\\n",
            "CA_001                              0.0                              0.0   \n",
            "CA_002                              0.0                              0.0   \n",
            "CA_003                              0.0                              0.0   \n",
            "CA_005                              0.0                              0.0   \n",
            "CA_007                              0.0                              0.0   \n",
            "\n",
            "        AATCTGAATCGCTATCTGAATCTGAATCGCT  AATCGCTATCCGAGTCTGAGTCGCTGTCTGA  ...  \\\n",
            "CA_001                              1.0                              0.0  ...   \n",
            "CA_002                              0.0                              0.0  ...   \n",
            "CA_003                              1.0                              0.0  ...   \n",
            "CA_005                              0.0                              0.0  ...   \n",
            "CA_007                              0.0                              0.0  ...   \n",
            "\n",
            "        GAATCTGAATCACTCGCTGAATCTGAACCAC  GAATCTGAATCACTCGCTGAGTCTGAATCAC  \\\n",
            "CA_001                              0.0                              0.0   \n",
            "CA_002                              0.0                              0.0   \n",
            "CA_003                              1.0                              0.0   \n",
            "CA_005                              0.0                              0.0   \n",
            "CA_007                              0.0                              0.0   \n",
            "\n",
            "        GAAGTGAGGCGCAAGCAGCAGAATTAGATAC  GATACCAAGCTTGTGAATTATTAACAAACTC  \\\n",
            "CA_001                              0.0                              0.0   \n",
            "CA_002                              0.0                              0.0   \n",
            "CA_003                              0.0                              0.0   \n",
            "CA_005                              0.0                              0.0   \n",
            "CA_007                              0.0                              0.0   \n",
            "\n",
            "        GAGTCTGAATCACTCGCTGAATCCGAGTCAC  GCAACTTTGACCATGCCCGCTTATATTGACC  \\\n",
            "CA_001                              0.0                              0.0   \n",
            "CA_002                              0.0                              0.0   \n",
            "CA_003                              0.0                              0.0   \n",
            "CA_005                              0.0                              0.0   \n",
            "CA_007                              0.0                              0.0   \n",
            "\n",
            "        GCTCTTGATACCAAGCTTGTGAATTATTAAC  GCTGAATCTGAATCACTCGCTGAATCTGAAC  \\\n",
            "CA_001                              0.0                              0.0   \n",
            "CA_002                              0.0                              0.0   \n",
            "CA_003                              0.0                              1.0   \n",
            "CA_005                              0.0                              0.0   \n",
            "CA_007                              0.0                              0.0   \n",
            "\n",
            "        GCGGTGTTGGAGTTTCTGGCTCACTCGGCAC  Label  \n",
            "CA_001                              0.0      1  \n",
            "CA_002                              0.0      1  \n",
            "CA_003                              0.0      0  \n",
            "CA_005                              0.0      1  \n",
            "CA_007                              0.0      0  \n",
            "\n",
            "[5 rows x 22436 columns]\n",
            "Kmer presence/absence dimensions:  (120, 22436)\n",
            "/content\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_4011a6f9-5fdc-4aff-8440-a8f5bb86913f\", \"GeneKmerMatrix.csv\", 11487807)"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_313961ed-55dc-48de-a21b-2202c8778e71\", \"GeneKmerMatrix.tsv\", 11487807)"
            ]
          },
          "metadata": {}
        }
      ],
      "source": [
        "# Initializing dataframe by adding the kmers of interest as column names\n",
        "df = pd.DataFrame(columns=kmers)\n",
        "\n",
        "# Getting path containing the genome fasta files of the 120 isolates in order to search the genome fasta files of each\n",
        "#... isolate for the kmers of interest.\n",
        "directory_path = os.path.join(os.getcwd(), \"MyDrive/ML Proj data/Isolates\")\n",
        "folder_content = os.listdir(directory_path)\n",
        "\n",
        "# Sorts the filenames stored in the folder_content list so that the isolates are added to the matrix in numerical order\n",
        "# ...and to ensure that rownames in the matrix have the same order as the labels dict\n",
        "folder_content=sorted(folder_content)\n",
        "print('Directory content:', folder_content)\n",
        "\n",
        "num=1\n",
        "print('Number\\t','Isolate ID\\t','Number of kmers found in isolate')\n",
        "\n",
        "for file in folder_content:\n",
        "\n",
        "  # Each loop reads 1 genome fasta file\n",
        "  isolate_name = file[:file.find('.txt')]\n",
        "  file_path = os.path.join(directory_path, file)\n",
        "  txt = open(file_path, 'r')\n",
        "\n",
        "  # For each file, a kmer_dict will be created which will store only the kmers from the genes of interest that were\n",
        "  # ...found in the genome fasta file of the current loop.\n",
        "  kmer_dict={} # Key: kmer found in genome fasta file, Value: 1 to signify presence\n",
        "\n",
        "  # Reading the kmers of current genome fasta file\n",
        "  for line in txt.readlines():\n",
        "    line = line.rstrip().split()\n",
        "\n",
        "    # A kmer from the current genome fasta file is only added to the kmer_dict if it is present in the dataframe columns\n",
        "    if line[0] in df.columns:\n",
        "      kmer_dict[line[0]] = 1\n",
        "  print(f'{num}' + '\\t' + f'  {isolate_name}' + '\\t\\t\\t' + f'{len(kmer_dict.keys())}')\n",
        "\n",
        "  # The kmer dict is used to add the current isolate to the dataframe as a new row\n",
        "  df.loc[isolate_name]=kmer_dict\n",
        "\n",
        "  # For all the kmer columns that were not present in the current isolate, these null cells will be filled with 0s\n",
        "  # ...to signify absence\n",
        "  df=df.fillna(0)\n",
        "  num+=1\n",
        "\n",
        "# Adding isolate IDs as rownames\n",
        "df.index=labels.keys()\n",
        "\n",
        "# Adding the 'Label' column\n",
        "df['Label'] = labels.values()\n",
        "\n",
        "# Making sure all empty cells are filled with zeros\n",
        "df=df.fillna(0)\n",
        "\n",
        "# Checking that there are no null values\n",
        "print('Are there any null values in the dataframe?', df.isnull().values.any())\n",
        "\n",
        "# Checking the dataframe has the correct format, rownames, column names, labels and dimensions\n",
        "print('Final Kmer presence/absence matrix:\\n', df.head())\n",
        "print('Kmer presence/absence dimensions: ', df.shape)\n",
        "\n",
        "# Writing the kmer presence/absence matrix to files and downloading these files to save the matrix\n",
        "from google.colab import files\n",
        "\n",
        "%cd /content\n",
        "df.to_csv('GeneKmerMatrix.csv', index=True)\n",
        "df.to_csv('GeneKmerMatrix.tsv', sep='\\t', index=True)\n",
        "\n",
        "files.download('GeneKmerMatrix.csv')\n",
        "files.download('GeneKmerMatrix.tsv')"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}